File: DAA/fractional-knapsack.cpp
#include <bits/stdc++.h>
using namespace std;

struct Item {
	int value;
	int weight;
};

bool comparator(Item a, Item b) {
	double x = (double)a.value/a.weight;
	double y = (double)b.value/b.weight;
	return x>y;
}

double maximumProfit(vector<Item>& items, int capacity) {
	sort(items.begin(), items.end(), comparator);
	double totalValue = 0.0;
	int currentWeight = 0;

	for(auto item: items) {
		if (currentWeight + item.weight <= capacity) {
			currentWeight += item.weight;
			totalValue += item.value;
		} else {
			int remain = capacity - currentWeight;
			totalValue += item.value * (double)remain/item.weight;
			break;
		}
	}
	return totalValue;
}

int main() {
	int n, capacity; 
	cin >> n >> capacity;
	vector<Item> items(n);
	for(auto& item: items)
		cin >> item.value;
	for(auto& item: items)
		cin >> item.weight;

	double result = maximumProfit(items, capacity);
	cout << result << "\n";

	return 0;
}

File: DAA/PYTHON/fractional-knapsack.py

class Item:
  def __init__(self, value, weight):
    self.value = value
    self.weight = weight
    self.cost = value // weight
  def __lt__(self, other):  
    return self.cost < other.cost 

def fractionalKnapsack(values, weights, capacity):
  items = []
  for i in range(len(values)):
    items.append(Item(values[i], weights[i]))
  
  items.sort(reverse=True)

  currentWeight = 0
  totalValue = 0.0

  for item in items:
    if currentWeight + item.weight <= capacity:
      currentWeight += item.weight
      totalValue += item.value
    else:
      remain = capacity - currentWeight
      totalValue = totalValue + item.value * float(remain/item.weight)
      break
  
  return totalValue 

n, capacity = 3, 50
values = [60, 100, 120]
weights = [10, 20, 30] 

result = fractionalKnapsack(values, weights, capacity)
print(result)

